### PR TITLE
some refactor and cleanup

### PR DESCRIPTION
**Pull request by djeebus**
make it clear which router supports the API publicly, vs implementation and monitoring details.
also split up some tests, add some around health check to prove it was double encoding, helm chart changes, etc

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -4,61 +4,59 @@ import (
...
+	agg := newAggregate()
 
-	// Serve metric endpoint
-	go func() {
-		log.Printf("metrics listening at %s", *metricsListen)
-		if err := http.ListenAndServe(*metricsListen, promhttp.Handler()); err != nil {
-			log.Panicf("error while serving metrics: %s", err)
-		}
-	}()
+	if apiListen != nil {
```
Comment (mplachter) [inline main.go:46]: this would never be `nil` as it has a `default` value in the flag

File: main.go [hunk 2]
```diff
@@ -4,61 +4,59 @@ import (
...
+		apiRouter := setupAPIRouter(corsDomain, agg)
+		go runServer("api", apiRouter, *apiListen)
+	}
+
+	if lifecycleListen != nil {
```
Comment (mplachter) [inline main.go:51]: this would never be `nil` as it has a `default` value in the flag

File: privateRouter.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	r.GET("/metrics", convertHandler(promhttp.Handler()))
+
+	return r
+}
+
+func convertHandler(h http.Handler) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		h.ServeHTTP(c.Writer, c.Request)
+	}
+}
```
Comment (mplachter) [inline privateRouter.go:24]: This should probably be moved to another `.go` file.

File: main.go [hunk 4]
```diff
@@ -4,61 +4,59 @@ import (
...
 	sigChannel := make(chan os.Signal, 1)
 	signal.Notify(sigChannel, syscall.SIGTERM, syscall.SIGINT)
 
-	rc := &RouterConfig{
-		AllowedCORS: cors,
+	runServer := func(label string, r *gin.Engine, listen string) {
+		log.Printf("%s server listening at %s", label, listen)
+		if err := r.Run(listen); err != nil {
+			log.Panicf("error while serving %s: %v", label, err)
+		}
```
Comment (mplachter) [inline main.go:41]: Can we move this into its own func declaration?

File: aggregate.go [hunk 5]
```diff
@@ -129,7 +131,7 @@ func (a *aggregate) parseAndMerge(r io.Reader, job string) error {
 	return nil
 }
 
-func (a *aggregate) handler(c *gin.Context) {
+func (a *aggregate) handleRender(c *gin.Context) {
```
Comment (mplachter) [inline aggregate.go:134]: Don't love the name but can't think of anything better :(

File: aggregate.go [hunk 6]
```diff
@@ -153,3 +155,20 @@ func (a *aggregate) handler(c *gin.Context) {
 
 	// TODO reset gauges
 }
+
+func (a *aggregate) handleInsert(c *gin.Context) {
```
Comment (mplachter) [inline aggregate.go:159]: Don't love the name but can't think of anything better :(

File: aggregate_test.go [hunk 7]
```diff
@@ -155,62 +156,54 @@ counter{a="a",b="b",job="test"} 3
...
-		{"labelFields", labelFields1, labelFields2, labelFieldResult, []string{}, nil, nil},
-		{"duplicateLabels", duplicateLabels, "", "", []string{}, fmt.Errorf("%s", duplicateError), nil},
-		{"reorderedLabels", reorderedLabels1, reorderedLabels2, reorderedLabelsResult, []string{}, nil, nil},
-		{"ignoredLabels", ignoredLabels1, ignoredLabels2, ignoredLabelsResult, []string{"ignore_me"}, nil, nil},
+		{"simpleGauge", gaugeInput, gaugeInput, gaugeOutput, []string{}},
+		{"in", in1, in2, want, []string{}},
+		{"multilabel", multilabel1, multilabel2, multilabelResult, []string{"ignore_label"}},
+		{"labelFields", labelFields1, labelFields2, labelFieldResult, []string{}},
+		{"reorderedLabels", reorderedLabels1, reorderedLabels2, reorderedLabelsResult, []string{}},
+		{"ignoredLabels", ignoredLabels1, ignoredLabels2, ignoredLabelsResult, []string{"ignore_me"}},
```
Comment (mplachter) [inline aggregate_test.go:171]: Please add the error test back in.

**Review by mplachter — COMMENTED** [review]
Some comments added

File: privateRouter.go (hunk 3 above)
Comment (djeebus) [reply privateRouter.go:24]: It's only used in privateRouter; makes more sense to keep it as close to the only place it's being used as possible, right?

File: aggregate.go (hunk 5 above)
Comment (djeebus) [reply aggregate.go:134]: yeah, agree. I think the weird thing is that it's a web handler inside of our "business" logic; I suspect this name would be less weird if we saw it in a router than in our datastore.

File: aggregate_test.go (hunk 7 above)
Comment (djeebus) [reply aggregate_test.go:171]: I moved it to its own `t.Run` below this one; it's a 3 line test that added 10 lines of complexity (give or take) to this test, so it seemed simpler to separate them.

File: privateRouter.go (hunk 3 above)
Comment (mplachter) [reply privateRouter.go:24]: True. Than I think moving into a `single routers.go` makes the most sense.

File: aggregate_test.go (hunk 7 above)
Comment (mplachter) [reply aggregate_test.go:171]: There are more `error` edge test cases we need to test for tho. Those tests are just not 100p completed yet.

File: aggregate.go [hunk 8]
```diff
@@ -137,25 +137,34 @@ func (a *aggregate) handleRender(c *gin.Context) {
 	enc := expfmt.NewEncoder(c.Writer, contentType)
 
 	a.familiesLock.RLock()
+	defer a.familiesLock.RUnlock()
```
Comment (mplachter) [inline aggregate.go:140]: This can't be used with flow control.

**Comment by djeebus** [comment]
```
[main][~/code/other/prom-aggregation-gateway]$ go test -benchmem -run=^$ -bench ^BenchmarkConcurrentAggregate$ prom-aggregation-gateway -benchtime=10s
goos: linux
goarch: amd64
pkg: prom-aggregation-gateway
cpu: AMD Ryzen 9 5900X 12-Core Processor            
BenchmarkConcurrentAggregate/metric_type_simpleGauge-24                   120742            100521 ns/op           90712 B/op       1403 allocs/op
BenchmarkConcurrentAggregate/metric_type_fullMetrics-24                    71344            173191 ns/op          197539 B/op       3893 allocs/op
BenchmarkConcurrentAggregate/metric_type_multiLabel-24                    189332             69238 ns/op           65427 B/op        693 allocs/op
BenchmarkConcurrentAggregate/metric_type_multiLabelIgnore-24              160431             66936 ns/op           66147 B/op        733 allocs/op
BenchmarkConcurrentAggregate/metric_type_labelFields-24                   197664             68502 ns/op           62792 B/op        583 allocs/op
BenchmarkConcurrentAggregate/metric_type_reorderedLabels-24               202158             63072 ns/op           64147 B/op        633 allocs/op
BenchmarkConcurrentAggregate/metric_type_ignoredLabels-24                 192980             70184 ns/op           65988 B/op        733 allocs/op
PASS
ok      prom-aggregation-gateway        94.144s
[clean-up][~/code/other/prom-aggregation-gateway]$ go test -benchmem -run=^$ -bench ^BenchmarkConcurrentAggregate$ prom-aggregation-gateway -benchtime=10s
goos: linux
goarch: amd64
pkg: prom-aggregation-gateway
cpu: AMD Ryzen 9 5900X 12-Core Processor            
BenchmarkConcurrentAggregate/metric_type_simpleGauge-24                   125487             92524 ns/op           90711 B/op       1403 allocs/op
BenchmarkConcurrentAggregate/metric_type_fullMetrics-24                    72567            165873 ns/op          197537 B/op       3893 allocs/op
BenchmarkConcurrentAggregate/metric_type_multiLabel-24                    184720             65368 ns/op           65427 B/op        693 allocs/op
BenchmarkConcurrentAggregate/metric_type_multiLabelIgnore-24              172918             68257 ns/op           66147 B/op        733 allocs/op
BenchmarkConcurrentAggregate/metric_type_labelFields-24                   180250             70082 ns/op           62792 B/op        583 allocs/op
BenchmarkConcurrentAggregate/metric_type_reorderedLabels-24               174849             69160 ns/op           64147 B/op        633 allocs/op
BenchmarkConcurrentAggregate/metric_type_ignoredLabels-24                 154258             71448 ns/op           65988 B/op        733 allocs/op
PASS
ok      prom-aggregation-gateway        89.447s```

**Review by mplachter — APPROVED** [review]
🚀

**Pull request merged**